### DISCUSSION TITLE
Minimal(or ideal) hardware requirements.

### DISCUSSION THREAD

**Discussion by candidodmv** [QuestPDF/QuestPDF#783] [General]
Hello everyone!

I fell missing the information about the hardware requirements, I searched for in general web also in the documentation, and found no thread about this topic.

I would like to know the personal experience of what is the "Ideal hardware requirements" to run QuestPdf with an acceptable time of response for a medium page size for a given report, such as 1000 pages.

Thanks in advance.

**Comment by girlpunk** [comment]
1000 pages is quite long, but that said the resource requirements for QuestPDF aren't particularly demanding.
Most of the library is single-threaded, so a high number of CPU cores won't particularly help, but a reasonable clock speed will (anything 2.5-3.0 Ghz is likely fine).
For RAM, you will need enough to comfortably hold all the layout information and resources, as well as the output file (possibly twice depending on how your application buffers it before sending/saving). This is unlikely to add up to more than a few hundred MB.

On top of that would be any requirements your program has when generating the content, layout, etc, which could be significantly higher, however that's not something anyone else can really speculate on.

**Reply by girlpunk** [reply]
For comparison, one of my test documents includes every feature in my program and produces a 250 page document (which would be considered very long in this industry) with lots of images, tables, etc. This takes about 15 seconds on my laptop with a 2.8Ghz i7 and 32GB RAM, though a lot of that is probably content registration and loading, rather than QuestPDF actually doing the layout/rendering.

**Comment by MarcinZiabek** [comment]
It all depends on the content of your document. 

For example, in one of the performance benchmarks, QuestPDF generates a 450 page long ebook in around 600ms on a MacBook with the Apple M2 processor. In another benchmark, a table with 1 million cells spanning nearly 3000 pages, takes around 30 seconds on the same machine.

On the other hand, if your document consists of many images that need to be scaled and compressed, the generation time may be much longer and demanding. @girlpunk presented this problem very accurately - the required RAM is roughly double/triple the size of your per-document resources. 

The library can be used in a non-production environment without any limitations, so feel free to create a simple document with only the heaviest assets inside (e.g. images only), and then evaluate the process in the Release mode.

**Comment by candidodmv** [comment]
Thanks @girlpunk and @MarcinZiabek  for share your insight.

As I supposed the performance it will depend of the type of the content that is being generated, proportionally more complexy content more time generating is required.

The document that I'm testing is with few images(two or less) but has a complex layout, that includes several cols and rows and tables with very much cols and rows(pure text).

It's taking a very much time(some minutes) which is not an user acceptable time. I will try increase the hardware of the host and evaluate the performance improvements.

Thanks.

**Reply by MarcinZiabek** [reply]
Please feel free to share a sample code or screenshot of the layout. This will help to provide some performance hints. What types of images are you using, in what form?

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
